src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_ObsoleteAttribute.vb

```diff
@@ -1146,6 +1146,156 @@ End Class
+    End Class
+End Namespace
+
+]]>
+    </file>
+</compilation>
+
+            Dim comp = CreateCompilationWithMscorlib40(source)
+            Dim diags = comp.GetDiagnostics()
+            diags.Verify(Diagnostic("TEST1", "M1()").WithArguments("Public Sub M1()").WithLocation(9, 9))
+
+            Dim diag = diags.Single()
+            Assert.Equal("TEST2", diag.Descriptor.HelpLinkUri)
+        End Sub
+
+        <Fact>
+        Public Sub TestObsoleteAttributeCustomPropertiesFromMetadata()
+            Dim source1 =
+<compilation>
+    <file name="a.vb"><![CDATA[Imports System
+
+Public Class C1
+    <Obsolete(DiagnosticId:="TEST1", UrlFormat:="TEST2")>
+    Sub M1()
+    End Sub
+End Class
+
+Namespace System
+    Public Class ObsoleteAttribute
+        Inherits Attribute
+    
+        Public Sub New()
+        End Sub
+    
+        Public Sub New(message As String)
+        End Sub
+    
+        Public Sub New(message As String, isError As Boolean)
+        End Sub
+    
+        Public Property DiagnosticId As String
+        Public Property UrlFormat As String
+    End Class
+End Namespace
+
+]]>
+    </file>
+</compilation>
+
+            Dim source2 =
+<compilation>
+    <file name="b.vb"><![CDATA[
+Class C2
+    Inherits C1
+    Sub M2()
+        M1()
+    End Sub
+End Class
+]]>
+    </file>
+</compilation>
+
+            Dim comp1 = CreateCompilationWithMscorlib40(source1)
+            comp1.VerifyDiagnostics()
+
+            Dim comp2 = CreateCompilationWithMscorlib40(source2, references:={comp1.EmitToImageReference()})
+            Dim diags = comp2.GetDiagnostics()
+            diags.Verify(Diagnostic("TEST1", "M1()").WithArguments("Public Sub M1()").WithLocation(4, 9))
+
+            Dim diag = diags.Single()
+            Assert.Equal("TEST2", diag.Descriptor.HelpLinkUri)
+        End Sub
+
+        <Fact>
+        Public Sub TestObsoleteAttributeDuplicateProperties()
+            Dim source =
+<compilation>
+    <file name="a.vb"><![CDATA[Imports System
+
+Class C1
+    <Obsolete(DiagnosticId:="TEST1", DiagnosticId:="TEST2", UrlFormat:="TEST3", UrlFormat:="TEST4")>
+    Sub M1()
+    End Sub
+
+    Sub M2()
+        M1()
+    End Sub
+End Class
+
+Namespace System
+    Public Class ObsoleteAttribute
+        Inherits Attribute
+    
+        Public Sub New()
+        End Sub
+    
+        Public Sub New(message As String)
+        End Sub
+    
+        Public Sub New(message As String, isError As Boolean)
+        End Sub
+    
+        Public Property DiagnosticId As String
+        Public Property UrlFormat As String
+    End Class
+End Namespace
+
+]]>
+    </file>
+</compilation>
+
+            Dim comp = CreateCompilationWithMscorlib40(source)
+            Dim diags = comp.GetDiagnostics()
+            diags.Verify(Diagnostic("TEST1", "M1()").WithArguments("Public Sub M1()").WithLocation(9, 9))
+
+            Dim diag = diags.Single()
+            Assert.Equal("TEST3", diag.Descriptor.HelpLinkUri)
+        End Sub
```

Comment:
>End Sub [](start = 8, length = 7)

Do you intend to clone all tests you added for C#?

Reply:
I was thinking I would copy a decent subset to ensure that we respect all combinations of the new properties from both source and metadata, as well as when different constructors are in use, and check handling of error scenarios in source. Do you think I should just clone all of them?

Reply:
> Do you think I should just clone all of them?

Unless some of them are redundant, I think cloning new tests is reasonable. 

---
In reply to: [399518710](https://github.com/dotnet/roslyn/pull/42518#discussion_r399518710) [](ancestors = 399518710)